### DISCUSSION TITLE
Formily2.0更新概要

### DISCUSSION THREAD

**Discussion by janryWang** [alibaba/formily#1087] [Ideas]
### 目标

- 进一步提升表单性能
- 解决大多数模棱两可的概念不清晰问题，降低学习成本

### 概要

Formily从uform0.x到至今差不多已经发布了2年多了，在这2年多的时间里，我们尽可能的做到了每次版本迭代都是向后兼容的，我也希望Formily可以做到永远向后兼容，可惜这个太难做到了，越来越发现这2年内堆了很多历史包袱，每次有用户提到一些边界问题的时候，我本人也很无奈去解决，因为要去解决A问题，很有可能就会对某个历史包袱产生互斥效果。所以，Formily2.0，只能忍痛重构，这个版本可以说是完全不一样的版本！可以先大概看一下示例代码(源码已推上github https://github.com/alibaba/formily/tree/formily_next )：

```tsx
import React, { useMemo } from 'react'
import {
  createForm,
  createSchemaField,
  FormProvider,
  FormConsumer,
  Field,
  ArrayField,
  onFieldReact,
  connect,
  mapProps,
  mapReadPretty,
  isVoidField,
} from '@formily/react'
import { Form, Input as AntdInput, Select, Card, Button, Space } from 'antd'
import 'antd/dist/antd.css'

const FormItem = connect(
  Form.Item,
  mapProps(
    { extract: 'validateStatus' },
    { extract: 'title', to: 'label' },
    (props, field) => {
      if (!field) return props
      if (isVoidField(field)) return props
      if (field.invalid) {
        return {
          help: field.editable ? field.errors : field.description,
        }
      } else {
        return {
          help: field.description,
        }
      }
    }
  )
)

const Input = connect(
  AntdInput,
  mapReadPretty(({ value }) => <span className="ant-form-text">{value}</span>)
)

const SchemaField = createSchemaField({
  components: {
    Card,
    Input,
    Select,
    FormItem,
  },
})

export default () => {
  const form = useMemo(
    () =>
      createForm({
        pattern: 'disabled',
        initialValues: {
          bb: '123',
        },
        effects: () => {
          onFieldReact('kk', (field, form) => {
            field.setDisplay(form.values.bb === '321' ? 'visibility' : 'none')
          })
          onFieldReact('aa.cc.*.dd', (field) => {
            const value = field.query('.ee').get('value')
            field.setPattern(value === '123' ? 'disabled' : 'editable')
          })
        },
      }),
    []
  )
  return (
    <FormProvider form={form}>
      <SchemaField>
        <SchemaField.String
          name="ioo"
          x-decorator="FormItem"
          x-decorator-props={{ tooltip: 'asdasd' }}
          title="IOO"
          x-component="Input"
          x-component-props={{ placeholder: 'asd' }}
        />
        <SchemaField.Void
          x-component="Card"
          x-component-props={{ style: { marginBottom: 10 } }}
        >
          <SchemaField.String
            name="poo"
            x-decorator="FormItem"
            title="POO"
            x-component="Input"
            x-reactions={[
              {
                dependencies: ['ioo'],
                when: '{{$deps[0] == 123}}',
                fullfill: {
                  state: {
                    value: '3333',
                  },
                },
                otherwise: {
                  state: {
                    value: '0000',
                  },
                },
              },
            ]}
          />
        </SchemaField.Void>
      </SchemaField>
      <Field
        name="bb"
        initialValue="123"
        required
        description={<div>122333</div>}
        title="BB"
        decorator={[FormItem, { tooltips: 'xxx', labelCol: { span: 3 } }]}
        component={[Input, { placeholder: 'xxx' }]}
        validator="url"
      />
      <Field name="kk" title="KK" required component={[Input]} />
      <Field name="aa">
        <ArrayField name="cc">
          {(field) => {
            const value = field.value || []
            return (
              <div>
                {value.map((item, index) => {
                  return (
                    <div key={item.id} style={{ marginBottom: 10 }}>
                      <Space>
                        <Field
                          name={`${index}.dd`}
                          required
                          component={[Input]}
                        />
                        <Field name={`${index}.ee`} component={[Input]} />
                        <Button
                          onClick={() => {
                            field.moveUp(index)
                          }}
                        >
                          上移
                        </Button>
                        <Button
                          onClick={() => {
                            field.moveDown(index)
                          }}
                        >
                          下移
                        </Button>
                        <Button
                          onClick={(e) => {
                            e.preventDefault()
                            field.remove(index)
                          }}
                        >
                          移除
                        </Button>
                      </Space>
                    </div>
                  )
                })}
                <Button
                  onClick={(e) => {
                    field.push({ id: new Date().getTime() })
                  }}
                >
                  添加
                </Button>
              </div>
            )
          }}
        </ArrayField>
      </Field>
      <FormConsumer>
        {(form) => {
          return (
            <div
              style={{
                background: '#eee',
                border: '1px solid #ddd',
                padding: 10,
                margin: '10px 0',
              }}
            >
              <code>
                <pre>
                  {JSON.stringify(form.query('aa').get('value'), null, 2)}
                </pre>
              </code>
            </div>
          )
        }}
      </FormConsumer>
      <FormConsumer>
        {(form) => {
          return (
            <div
              style={{
                background: '#eee',
                border: '1px solid #ddd',
                padding: 10,
                margin: '10px 0',
              }}
            >
              <code>
                <pre>{JSON.stringify(form.errors, null, 2)}</pre>
              </code>
            </div>
          )
        }}
      </FormConsumer>
      <Space>
        <Button
          onClick={() => {
            form.submit(console.log)
          }}
        >
          提交
        </Button>
        <Button
          onClick={() => {
            form.setPattern('editable')
          }}
        >
          编辑
        </Button>
        <Button
          onClick={() => {
            form.setPattern('disabled')
          }}
        >
          禁用
        </Button>
      </Space>
    </FormProvider>
  )
}
```

不多说了，下面来看具体有哪些改动点：

### Core层

1. 内核从immerjs完全迁移到了mobx6，为什么要用mobx？
   1. 因为mobx的依赖收集机制，远比过去用immer始终还是逃脱不了脏检查的性能要好。
   2. mobx非常适合构建领域充血模型，代码简洁度极高，可维护性也很高。
1. 新增Field模型的`display("none" | "visibility" | "hidden")`属性
    1. 相比过去的visible属性，更清晰，也和前端的自然联想比较贴切，display为none的时候，就是数据和UI都隐藏，display为hidden的时候，就是UI隐藏，数据不隐藏
1. 新增Field模型的`pattern("editable" | "disabled" | "readOnly" | "readPretty")`属性
    1. 相比过去的editable属性，它描述的是字段的交互设计模式形态，而非仅仅只是一个可编辑和非可编辑态，这也是在给用户答疑过程中遇到的问题，因为用户很难知道过去的editable属性到底是映射组件的disabled好呢，还是readOnly好呢，还是文本展示好呢。给用户的选择成本太高，所以，这样一改，完全可以做到非常清晰的映射组件的交互模式。
1. 移除Field模型的name属性，path属性含义改变为数据路径，新增address属性，代表字段地址
    1. 之前用name属性代表数据路径其实也是模棱两可，对于新用户而言完全不知道它到底是名称还是路径，所以移除
    2. path属性很明显的代表路径，所以就用来描述数据操作的路径。
    3. address属性代表地址，地址就是绝对路径，也是比较贴切的。
1. Field模型的rules属性改为validator属性，因为rules，太广泛，不能直接描述校验规则，所以改为validator属性，直接描述这个是校验器
2. Field模型的props属性含义变为Field的props，而非组件的扩展属性，新增decorator属性和component属性，用来表达字段的渲染组件和装饰器组件
    1. 过去的Field模型并没有把组件相关的属性做很好的区分，都是放到了state.props里，同样还是会带来一个理解上的歧义问题，这个props到底是谁的属性。所以，在这里定义decorator和component属性，而且他们的数据结构都是`[组件函数,组件属性]` 这样的形式
1. 校验反馈新增success类型，在过去Formily支持了errror类型和warning类型，为了提升模型完备性，新增了success类型
2. 内置effects处理逻辑，UI桥接层不再需要依赖react-eva，统一下沉到内核层来处理，核心也是为了提升内核的可移植性
3. Field模型values属性改为inputValues，同时新增inputValue，前者代表用户输入事件的多参值，后者代表用户输入事件的值
4. Field模型新增title/description属性，用来描述字段标题和描述，更加方便与json-schema做映射
5. Field模型新增dataSource属性，用来描述字段数据源，更加方便与json-schema的enum做映射，也方便与各种枚举型组件做映射
6. Form模型和Field模型新增query函数，专门用来做节点查询，路径查询语法支持按照相对路径查询节点
7. 新增ArrayField/ObjectField/VoidField模型，移除过去的VirtualField模型，为什么要把Virtual改为Void，因为Void更适合表达数据上的虚拟概念，比如js里的void关键字就代表不返回值
8. 内置Formily devtools交互，不再需要在UI桥阶层处理，同样是为了提升内核的可移植性
9. FormEffectHooks移除rxjs依赖，只是最简单的回调函数，同时用户直接import对应的EffectHook即可，不需要再从FormEffectHooks中导出，这样也方便用户使用了。

### React层

1. 移除对react-eva的依赖，也就是移除了对rxjs的依赖，体积会小很多。同时effects不再支持`$("onFieldChange")`这种写法，统一使用`onFieldValueChange('aa.bb',(field,form)=>{})`这样的写法
2. 不再支持受控更新，比如过去我们是可以靠全局重绘来更新Field的属性的，现在不再支持，想要更新属性，统一通过操作 `form.query('aa.bb').get(field=>field.setComponentProps(xxx))` 这样来更新状态，为什么不再支持受控更新，主要是这里的边界问题太多了，因为本身Formily主要是推荐非受控写法，如果又支持受控写法，那就会出现冲突问题。而且为了支持受控，之前还做了脏检查，又进一步影响了性能，所以，Field组件声明式写法的属性，只在初始化生效，如果要动态控制，就需要使用对应的API来做命令式控制。
3. Field组件路径自动继承，之前使用Field组件，传递的name属性必须是绝对路径，现在传递的name会自动继承父组件路径，如果需要改为绝对路径，则传入一个basePath来重置父路径
4. Field组件支持component属性和decorator属性，两个属性都是`[组件,组件属性]` 数组结构，这样传参可以获得更好的代码提示体验，同时也能和本身Field的属性做解耦
5. 移除Form组件，新增FormProvider组件，因为Formily的核心是负责做表单状态管理，而非管理UI，同时大多数UI组件库都有Form组件，Formily再提供一个Form组件，就必须要用户做桥接，倒不如直接移除Form组件，使用哪个组件库就用哪个Form组件即可。
6. FormProvider必须接收一个form属性，form属性需要通过createForm来创建form实例，如果不创建，则Formily跑不起来。这个步骤其实和过去的createFormActions类似，只不过现在整个Formily变成了强依赖外部传入的form实例，这样做的好处则是更加符合react哲学，组件都是木偶组件，状态外置管理。
7. 移除FormSpy组件，新增FormConsumer组件，专门用来响应式消费Form状态
8. 提供ArrayField/ObjectField/VoidField组件，专门用来维护不同数据类型的充血模型的
9. useForm和useField API只负责读取上下文中的form实例和field实例，不再负责创建Form和创建Field，在过去的版本中，这样的hook专门用于创建实例，有很多用户都会用错，很是歧义，所以现在行为需要转变
10. connect API下层到React层，过去connect都是在Schema层的，因为现在我们的模型能承载decorator和component了，所以connect自然需要在UI桥阶层来处理，当然，默认组件都不需要connect，只要组件支持了value/onChange，即可立马接入，只是如果你需要支持pattern或者其他状态映射，则需要手动包装connect。同时我们还提供了便捷函数mapProps和mapReadPretty来映射组件属性和优雅阅读态的组件
11. useFormQuery API暂时移除，后面考虑迁移到alist上去

### Schema层

1. react-schema-renderer移除，只提供了一个API，就是createSchemaField工厂函数,也就是没有SchemaForm组件了，同时把createSchemaField放到了@formily/react里
2. 借助工厂函数创建出来的SchemaField，可以导出一个Markup组件，Markup组件与之前的SchemaMarkupField组件等价，只是它现在拥有非常完备的代码提示能力，之前我们指定x-component的时候，x-component-props则无法获得智能提示，现在有了工厂函数了之后，我们则获得了非常智能的代码提示。
3. 移除了x-linkages，新增了x-reactions，过去的主动联动模式，对于简单联动而言成本还好，但是对于复杂一些的联动，代码量就会变得巨大，为了优化代码量，借助x-reactions可以实现被动联动，代码量可以大大减少。
4. 移除x-props，新增x-decorator和x-decorator-props，x-decorator我们可以指定任意组件，可以是FormItem，也可以是FormMegaItem，同时x-decorator-props就是对应的组件属性，也就是我们的JSON Schema描述表单，对于装饰器也可以动态灵活的描述。
5. 移除x-rules，替换为x-validator，与Field模型的属性保持一致
6. 新增x-pattern/x-display属性，与Field模型属性保持一致
7. JSON Schema的type如果指定为void，则会以VoidField来做状态承接，不会返回值
8. 协议层支持向后兼容

### 组件桥接层

1. 移除@formily/antd-components，统一收敛到@formily/antd中，用户无需引入两个包
2. 移除@formily/next-components，统一收敛到@formily/next中，用户无需引入两个包
3. 移除styled-components依赖，样式统一follow具体组件库，比如antd用的less，fusion用的scss

### 总结

Formily2.x的更新，变化非常巨大，基本上完全做不到平滑迁移，也没法用codemod工具做迁移，所以，需要用户非常谨慎的升级。但是它的收益还是比较高的，一个是性能大幅度提升，一个是学习成本会大大降低

**Comment by lushanyanYX** [comment]
首先感谢formily，每次提问题修复都很及时，代码质量也很高

我使用主要是schema模式，完全配置化的表单、或者导出schema并用formily渲染，外层写自己的业务逻辑

我比较困惑的问题有：
1. value、defaultValue、initialValues的区别没有很清楚，很多问题都是试出来的，schema中的default值也有一些问题，最后我们不得不自己写解析函数，把schema的默认值还有请求数据得到的值组装起来传进去，很绕
2. schema模式和jsx模式能力有差异，文档部分没有明确体现出来，比如我想实现一个复杂联动功能，自己测试才发现schema模式不支持，需要在schema模式通过写js去给做增强。是否考虑文档变成基础功能(同一个功能schema模式、jsx模式都可以使用)、进阶功能(只有jsx模式可以使用，然后给出schema模式自己实现类似的功能的最佳实践)
3. 组件的入参嵌套了很多层，然后使用useSchemaProps API看着里面那么多的参数也很迷惑

**Reply by janryWang** [reply]
schema模式的话，因为移除了x-linkages，联动这一块的写法得改成x-reactions，改造成本可能会比较大，像x-pattern和x-display这个前端层做这个转换还好

**Reply by janryWang** [reply]
1. 因为没有Form组件了，所以value/defaultValue被干掉了，只有createForm API，可以给它传入values和initialValues，然后schema中的default和initialValues是绝对幂等的，所以也不需要你再重新组装
2. schema模式和jsx模式在2.x是可以完美融合的，因为schema模式就只是在SchemaField这个组件中体现，两者都是共享上下文的，我可以在jsx的自定义组件里使用SchemaField，也可以在SchemaField的自定义组件里使用原生Field
3. 以后没有useSchemaProps了，只有useField和useForm两个hook，

**Reply by lushanyanYX** [reply]
> schema模式的话，因为移除了x-linkages，联动这一块的写法得改成x-reactions，改造成本可能会比较大，像x-pattern和x-display这个前端层做这个转换还好

改造没问题，之前的x-linkages和jsx的联动能实现的功能有差异，把配置折腾顺就很费劲了，比如arraycard的联动直接使用x-linkages感觉做不了
文档预设了使用schema模式和jsx模式的两类用户，但是实际上用schema的用户(比如我)，如果功能不满足，我就想怎么去在schema模式渲染基础上，通过jsx实现联动等功能

**Reply by janryWang** [reply]
现在要实现联动，schema模式下，就是x-reactions，jsx场景下，还是使用effects，不过不再需要走onFieldValueChange去控制了，直接用onFieldReact即可，这种被动联动模式是走的mobx依赖收集，性能会更好

**Comment by Lind-pro** [comment]
预计什么时候发布新版本呢？

**Reply by janryWang** [reply]
目前不太好预估，文档和单测是工作量巨大的，感觉还需要社区的帮助，一起共建

**Reply by Lind-pro** [reply]
> 目前不太好预估，文档和单测是工作量巨大的，感觉还需要社区的帮助，一起共建

那就是暂时还是先用老版本吗。。。。新项目

**Reply by janryWang** [reply]
是的，1.x还会持续维护一年吧，2.x还需要慢慢过渡的

**Reply by o-w-o** [reply]
> 目前不太好预估，文档和单测是工作量巨大的，感觉还需要社区的帮助，一起共建

官方准备啥时候发布 `2.0 bate` 版的 npm 包？可以先尝尝鲜 😄 ，新版本感觉简明了不少。

**Comment by liuweiGL** [comment]
api 层面的变动太大了，感觉目前手头上的项目基本上无法迁移到 2.x 了。项目中不能使用 2.x 的话，就不能很好的参与进来，因为很多东西（用法、自定义组件）都是实践总结出来的。

**Reply by janryWang** [reply]
可以在新项目中使用，或者找一些边缘页面，风险并不高的页面试水

**Comment by MatriXiao88** [comment]
`form.query('aa.bb').get(field=>field.setComponentProps(xxx))`
这样的调用链用起来还是显得复杂，从`query`的语义来理解，在`query`的第二个参数作为一个回调函数岂不是更直观，这里在`query`的返回对象再调用`get`，直观上有点重复
可以考虑这样：`form.query('aa.bb', (field)=>{})`？

**Reply by janryWang** [reply]
一开始是想这样处理的，但是因为字段模型是分了Field/ArrayField/ObjectField/VoidField，每种模型又有不同的方法，为了方便区分类型才使用链式调用的，比如`form.query('aa.bb').array.get(field=>{}) or form.query('aa.bb').void.get(field=>{}) `，这样就不需要用户用isArrayField/isObjectField这样的API来判断类型之后再调用方法

**Comment by nnecec** [comment]
没太看出来 schema 模式改动的大不大。因为前端的调整是可以评估的，但是 schema 在业务使用中，大部分存在了服务端，如果改动较大则不容易推动升级。或者是不是可以考虑像 antd4 的升级一样，对老版本单独发一个依赖包。

**Comment by yee94** [comment]
期待新版本发布

**Comment by umi-soft** [comment]
啥时候可以支持vue的生态呢？

**Reply by janryWang** [reply]
vue这块只能是社区来贡献，官方团队会提供技术支持，不会做深度建设，主要还是看社区对Formily的态度吧，目前已经有了support vue PR，还没合并，我联系PR作者直接以2.x来支持，他那边是愿意的，所以大概率会有@formily/vue这样的包，至于@formily/element这个包，还是得看看社区的态度

**Reply by umi-soft** [reply]
> vue这块只能是社区来贡献，官方团队会提供技术支持，不会做深度建设，主要还是看社区对Formily的态度吧，目前已经有了support vue PR，还没合并，我联系PR作者直接以2.x来支持，他那边是愿意的，所以大概率会有@formily/vue这样的包，至于@formily/element这个包，还是得看看社区的态度

想了解一下，【@formily/vue】进展如何呢？

**Comment by wsxe9988** [comment]
啥时候可以支持vue的生态呢？

**Comment by ChangeHow** [comment]
我比较疑惑的一点是，为什么要给setDisplay为none的fields的行为设置为删除？我个人有个疑惑：

**对于数据提交来说，这个视图上不存在节点是否应该不存在对应的字段（仅针对none来说）是否是必要和符合直觉的？（这是我非常疑惑的一点）**

个人以为，这里很容易出现一些问题和增加维护成本？比如怎么为这个可能动态添加和删除的字段增加监听器？删除前后触发effect会不会导致一些问题？ 🤕

**Reply by janryWang** [reply]
大多数场景UI上不显示字段了，字段的数据就不应该被提交，但是UI上不显示字段，但又要提交数据则在一些场景上很有必要，再看看display属性，为None的时候，在CSS中是UI不隐藏，同时在文档流中也隐藏，为Hidden的时候，则是UI隐藏，文档流不隐藏，恰好可以类比到表单字段上

**Reply by ChangeHow** [reply]
确实，类比HTML表现是一个很直观的想法，但是我最近在琢磨，字段这个东西，对于form来说，是否是客观存在的？我们即使是隐藏了UI，**同步修改**数据结构这个是否对于form来说是相当有必要的？或者这个步骤可以让使用者在submit中自己去处理他们。

我觉得可以考虑一个场景：当我们需要提交一些信息，而信息是根据勾选项动态渲染的一个动态表单，那么，这些数据，对于后端而言必然需要去列举出所有能够支持的表单场景，这时候，如果我们更进一步，做了属性的tree-shaking，那么，这部分的判空的成本是否是转嫁到了后端？

退一步，如果没有这一步的shaking，单单就使用来说，是否有很大的倒退？

*非常希望能够看到你的考量*

**Reply by janryWang** [reply]
我个人的理解，如果不支持同步修改数据，对于前端开发者而言，开发工作量巨大，因为本不存在属性tree-shaking一说，而且提交时候的数据完全是无上下文的纯json，它不知道，XX字段是在某次联动过程中被隐藏，同时不应该被提交，所以完全没法做到自动shaking，只能在onSubmit的时候，自己人肉去找到某个字段，然后delete掉，这种机械且低效的逻辑，想想就恶心

**Reply by ChangeHow** [reply]
懂了，感谢

**Comment by o-w-o** [comment]
提个小小的建议，formily 2.0 官网导航栏 logo 可以改成 `svg` 或 `png` 格式的。

![image](https://user-images.githubusercontent.com/12809186/102193813-85f1e980-3ef7-11eb-8100-b5cb6d7cfd06.png)

强迫症犯了 🌚

**Comment by hangaoke1** [comment]
不知道依赖这块是否还能做精简，写过1.0的表单以及看了这个2.0的demo，写一个表单需要import不少依赖，有一定的使用成本。

**Reply by janryWang** [reply]
没办法，对于外部依赖，如果不设置peerDependencies，就很容易重复打包，不知道各位有没有比较好的想法

**Comment by hangaoke1** [comment]
```tsx
const form = createForm({
  effects: () => {
    onFieldReact('input', (field, form) => {
      console.log('>>> textarea')
      const value = field.query('.textarea').get(f => {
        const textareaValue = f.value
      })
    })
  }
})
```
onFieldReact这个api设计感觉会有点争议，字面上看，像是对input匹配的field做响应式触发，实际上是回调里面所有使用到的filed都会被mobx收集，如上例子，操作textarea这个field就会触发回调。 理论上就会出现重复计算问题，比如price * count = total，这种场景，修改price或者count 会触发两次total的计算。

**Reply by janryWang** [reply]
这里是操作textarea字段的值且发生了数据变化才会触发回调，不会出现重复计算的，不过你说的onFieldReact感觉是受之前使用onFieldChange之后有先入为主的思路了，这个是需要习惯一下

**Comment by wittech** [comment]
测试了一下发现最新的2.0版本，表单input组件无法响应onFocus和onBlur事件，从field对象拿到的字段状态active一直是false。看了ReactiveField的实现代码，发现react的版本里面只有处理onChange事件，没有onFocus和onBlur的处理逻辑。而vue版本里面却是实现了的，请问是内核还没有完全实现？

**Comment by DophinL** [comment]
看完之后有很多收获！赞

**Comment by YangXiaochun-admin** [comment]
想问下@formily/vue什么时候会发布，目前项目中动态表单打算尝试下

**Reply by xfz1987** [reply]
直接给表单设置 form.setPattern('readOnly')，没有任何效果

**Comment by moweiwei** [comment]
请问 json-schema 协议用的是什么版本呢？

**Comment by caodonghai** [comment]
基于formily的动态表单什么支持服务端渲染（SSR）吗？
